### PR TITLE
Obsolete concurrency limiter middleware.

### PR DESCRIPTION
**Pull request by mitchdenny**
Addresses https://github.com/dotnet/aspnetcore/issues/47532

### CODE REVIEW

File: src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterExtensions.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ public static class ConcurrencyLimiterExtensions
     /// </summary>
     /// <param name="app">The <see cref="IApplicationBuilder"/>.</param>
     /// <returns>The <see cref="IApplicationBuilder"/>.</returns>
+    [Obsolete()]
```
Comment (davidfowl) [inline src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterExtensions.cs:18]: Why is this message empty?

File: src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.AspNetCore.ConcurrencyLimiter;
 /// <summary>
 /// Limits the number of concurrent requests allowed in the application.
 /// </summary>
+[Obsolete("Concurrency Limiter middleware has been deprecated, use new rate limiting middleware. See: aspnet/rate-limiting")]
```
Comment (JamesNK) [inline src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs:13]:
```suggestion
[Obsolete("Concurrency Limiter middleware has been deprecated and will be removed in a future release. Update the app to use concurrency features in rate limiting middleware. For more information, see https://aka.ms/aspnet/rate-limiting")]
```

aka.ms link goes to `https://learn.microsoft.com/aspnet/core/performance/rate-limit` (note: no en-us in the link so user is redirected based on browser culture).

Use on all obsoletes.

File: src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterExtensions.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterExtensions.cs:18]: It's a rebel. It has been crushed in one swift stroke.

**Comment by mitchdenny** [comment]
Adding blog candidate tag for this so we can let folks know that we are obsoleting this middleware.

**Comment by mgravell** [comment]
should the pragmas in here be switch to (or supplemented with) CS0618? currently triggering CI fail (noticed in build-ops rotation)

**Comment by mgravell** [comment]
see also src/Middleware/ConcurrencyLimiter/perf/Microbenchmarks/QueueEmptyOverhead.cs which is tripping CI failure

**Comment by Tratcher** [comment]
> - [ ] Create an announcement issue by using the [ASP.NET Core breaking change](https://github.com/aspnet/Announcements/issues/new?assignees=&labels=Breaking+change&template=breaking-change.yaml&title=%5BBreaking+change%5D%3A+) issue template.

Please draft the announcement here in a PR comment for review.

We likely don't need to mention this in the blog, it's low usage.

**Review by JamesNK — APPROVED** [review]
LGTM. Remember to add a breaking change notification.

**Comment by mitchdenny** [comment]
@Tratcher proposed wording for announcement:

> In .NET 8.0 the ASP.NET Core team is deprecating the `ConcurrencyLimiterMiddleware` and associated methods and types (e.g. `UseConcurrencyLimiter(...)`). This package will be removed in .NET 9.0.
> 
> Developers requiring rate limiting capabilties should switch to the newer and more capable rate limiting middleware that was introduced in .NET 7.0 (e.g. `UseRateLimiter(...)`). The .NET 7.0 rate limiting API includes a concurrency limiter along with several other rate limiting algorithms that you can apply to your application.
> 
> For more information on rate limiting in ASP.NET Core see:
> https://learn.microsoft.com/aspnet/core/performance/rate-limit

**Comment by JamesNK** [comment]
That looks good for a summary. Here is an example of obsoleting an API announcement: https://github.com/aspnet/Announcements/issues/458

**Comment by gewarren** [comment]
Is there a reason ASP.NET doesn't use custom obsoletion IDs like core .NET libraries? https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/obsoletions-overview

**Comment by mitchdenny** [comment]
/cc @davidfowl for his answers to this:

> Is there a reason ASP.NET doesn't use custom obsoletion IDs like core .NET libraries? https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/obsoletions-overview

**Comment by Youssef1313** [comment]
Removing needs-breaking-change-announcement.

This was announced in https://github.com/aspnet/Announcements/issues/502

**Pull request merged**